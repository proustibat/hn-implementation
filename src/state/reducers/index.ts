import { combineReducers } from 'redux';
import stories, {StoriesReducerType} from './storiesReducer';
export type RootReducersType = {
    stories: StoriesReducerType
};

export default combineReducers({
  stories,
});
