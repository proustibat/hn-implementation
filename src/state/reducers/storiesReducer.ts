import { storiesActionTypes as Actions } from '../actionsTypes';
import { AxiosResponse } from 'axios';

export enum StoryKindType {
  job = 'job',
  story = 'story',
  comment = 'comment',
  poll = 'poll',
  pollopt = 'pollopt',
}

export type StoryProps = {
  id: number;
  deleted?: boolean; //	true if the item is deleted.
  type?: StoryKindType; // The type of item.
  by?: string; // The username of the item's author.
  time?: number; //	Creation date of the item, in Unix Time.
  text?: string; //	The comment, story or poll text. HTML.
  dead?: boolean; // true if the item is dead.
  parent?: number; // The comment's parent: either another comment or the relevant story.
  poll?: number; //	The pollopt's associated poll.
  kids?: number[]; // The ids of the item's comments, in ranked display order.
  url?: string; //	The URL of the story.
  score?: number; // The story's score, or the votes for a pollopt.
  title?: string; // The title of the story, poll or job. HTML.
  parts?: number[]; // A list of related pollopts, in display order.
  descendants?: number; //	In the case of stories or polls, the total comment count.
  loading?: boolean;
  loaded?: boolean;
};

export type StoriesReducerType = {
  items: StoryProps[];
  loading: boolean;
  loaded: boolean;
};

export const initialState: StoriesReducerType = {
  items: [],
  loading: false,
  loaded: false,
};

const LIMIT = 10;

const storiesReducer = (
  state = initialState,
  action: {
    type: Actions;
    payload?: { data?: unknown; request?: AxiosResponse };
  },
) => {
  let loadedItemId: number;
  switch (action.type) {
    case Actions.REQUEST_STORIES:
      return {
        ...state,
        loading: true,
        loaded: false,
      };
    case Actions.REQUEST_STORIES_FAILURE:
      return {
        ...state,
        loading: false,
        loaded: false,
      };
    case Actions.REQUEST_STORIES_SUCCESS:
      return {
        ...state,
        // slice temp while we don't manage queue or pagination
        items: (action?.payload?.data as number[]).slice(0, LIMIT).map(id => ({
          id,
          loaded: false,
          loading: false,
        })) as StoryProps[],
        loading: false,
        loaded: true,
      };
    case Actions.REQUEST_STORY:
      loadedItemId = action.payload?.request?.data?.id as number;
      return {
        ...state,
        items: state.items.map(item => {
          return item.id === loadedItemId
            ? {
                ...item,
                loading: true,
                loaded: false,
              }
            : item;
        }),
      };
    case Actions.REQUEST_STORY_FAILURE:
      loadedItemId = action.payload?.request?.data?.id as number;
      return {
        ...state,
        items: state.items.map(item => {
          return item.id === loadedItemId
            ? {
                ...item,
                loading: false,
                loaded: false,
              }
            : item;
        }),
      };
    case Actions.REQUEST_STORY_SUCCESS:
      const loadedItem = action.payload?.data as StoryProps;
      return {
        ...state,
        items: state.items.map(item => {
          return loadedItem && item.id === loadedItem.id
            ? {
                ...item,
                ...loadedItem,
                loading: false,
                loaded: true,
              }
            : item;
        }),
      };
    default:
      return state;
  }
};

export default storiesReducer;
