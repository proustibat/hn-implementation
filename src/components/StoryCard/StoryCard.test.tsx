import React from 'react';
import { render } from '@testing-library/react';
import { StoryCard } from './index';
import { StoryKindType, StoryProps } from '../../state/reducers/storiesReducer';

const props: StoryProps = {
  id: new Date().getTime(),
  title: 'my-title',
  score: 11,
  loaded: true,
  url: 'https://www.werkspot.nl/',
  text: 'text',
};

describe('StoryCard', () => {
  const types = [
    [StoryKindType.story, StoryKindType.story],
    [StoryKindType.comment, StoryKindType.comment],
    [StoryKindType.job, StoryKindType.job],
    [StoryKindType.poll, StoryKindType.poll],
    [StoryKindType.pollopt, StoryKindType.pollopt],
  ];

  it.each(types)('should render the StoryCard with %s type', (...args: any) => {
    const [, type] = args;

    // Given / When
    const { baseElement } = render(<StoryCard {...props} type={type} />);

    // Then
    expect(baseElement).toMatchSnapshot();
  });

  it('should not render anything if data are not loaded', () => {
    // Given / When
    const { baseElement } = render(
      <StoryCard {...props} type={StoryKindType.story} loaded={false} />,
    );

    // Then
    expect(baseElement).toMatchSnapshot();
  });

  it('should render loader', () => {
    // Given / When
    const { baseElement } = render(
      <StoryCard {...props} type={StoryKindType.story} loading={true} />,
    );

    // Then
    expect(baseElement).toMatchSnapshot();
  });
});
