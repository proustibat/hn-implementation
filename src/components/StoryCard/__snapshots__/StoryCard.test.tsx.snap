// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryCard should not render anything if data are not loaded 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiCard-root rootNoLoaded_padding-0 rootNoLoaded_margin-0 MuiPaper-elevation1 MuiPaper-rounded"
    />
  </div>
</body>
`;

exports[`StoryCard should render loader 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
        role="progressbar"
      >
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
        />
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`StoryCard should render the StoryCard with comment type 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4-.01-18zM18 14H6v-2h12v2zm0-3H6V9h12v2zm0-3H6V6h12v2z"
            />
          </svg>
          <span
            class="MuiChip-label MuiChip-labelSmall"
          >
            comment
          </span>
        </div>
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          my-title
        </h2>
        <p
          class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          score
          : 
          11
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          text
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          href="https://www.werkspot.nl/"
          tabindex="0"
          target="_blank"
        >
          <span
            class="MuiButton-label"
          >
            visit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </div>
    </div>
  </div>
</body>
`;

exports[`StoryCard should render the StoryCard with job type 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 6h-4V4c0-1.11-.89-2-2-2h-4c-1.11 0-2 .89-2 2v2H4c-1.11 0-1.99.89-1.99 2L2 19c0 1.11.89 2 2 2h16c1.11 0 2-.89 2-2V8c0-1.11-.89-2-2-2zm-6 0h-4V4h4v2z"
            />
          </svg>
          <span
            class="MuiChip-label MuiChip-labelSmall"
          >
            job
          </span>
        </div>
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          my-title
        </h2>
        <p
          class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          score
          : 
          11
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          text
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          href="https://www.werkspot.nl/"
          tabindex="0"
          target="_blank"
        >
          <span
            class="MuiButton-label"
          >
            visit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </div>
    </div>
  </div>
</body>
`;

exports[`StoryCard should render the StoryCard with poll type 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z"
            />
          </svg>
          <span
            class="MuiChip-label MuiChip-labelSmall"
          >
            poll
          </span>
        </div>
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          my-title
        </h2>
        <p
          class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          score
          : 
          11
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          text
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          href="https://www.werkspot.nl/"
          tabindex="0"
          target="_blank"
        >
          <span
            class="MuiButton-label"
          >
            visit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </div>
    </div>
  </div>
</body>
`;

exports[`StoryCard should render the StoryCard with pollopt type 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
          <span
            class="MuiChip-label MuiChip-labelSmall"
          >
            pollopt
          </span>
        </div>
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          my-title
        </h2>
        <p
          class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          score
          : 
          11
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          text
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          href="https://www.werkspot.nl/"
          tabindex="0"
          target="_blank"
        >
          <span
            class="MuiButton-label"
          >
            visit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </div>
    </div>
  </div>
</body>
`;

exports[`StoryCard should render the StoryCard with story type 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M14 17H4v2h10v-2zm6-8H4v2h16V9zM4 15h16v-2H4v2zM4 5v2h16V5H4z"
            />
          </svg>
          <span
            class="MuiChip-label MuiChip-labelSmall"
          >
            story
          </span>
        </div>
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          my-title
        </h2>
        <p
          class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          score
          : 
          11
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          text
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          href="https://www.werkspot.nl/"
          tabindex="0"
          target="_blank"
        >
          <span
            class="MuiButton-label"
          >
            visit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </div>
    </div>
  </div>
</body>
`;
